pkg/storegateway/bucket_index_postings.go

```diff
@@ -322,3 +324,143 @@ func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
+	return groups, nil
+}
+
+// worstCaseFetchedDataStrategy select a few of the posting groups such that their total size
+// does not exceed the size of series in the worst case. The worst case is fetching all series
+// in the smallest non-subtractive posting group - this is effectively the
+// upper bound on how many series all the posting groups can select after being intersected/subtracted.
+//
+// This strategy is meant to prevent fetching some the largest posting lists in an index. Those are usually
+// less selective the smaller ones and only add cost to fetching.
+//
+// The strategy greedily selects the first N posting groups (sorted in ascending order by their size)
+// whose combined size doesn't exceed the size of series in the worst case. The rest of the posting groups
+// are omitted.
+// worstCaseFetchedDataStrategy uses a fixed estimation about the size of series in the index (tsdb.EstimatedSeriesP99Size).
+type worstCaseFetchedDataStrategy struct {
+	// postingListActualSizeFactor affects how
+	// posting lists are summed together.
+	// This factor is multiplied by the size of posting lists
+	// in the block index.
+	postingListActualSizeFactor float64
+}
+
+func (s worstCaseFetchedDataStrategy) name() string {
+	return fmt.Sprintf("worstCase%0.1f", s.postingListActualSizeFactor)
+}
+
+func (s worstCaseFetchedDataStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	const seriesBytesPerPostingByte = tsdb.EstimatedSeriesP99Size / tsdb.BytesPerPostingInAPostingList
+
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].totalSize < groups[j].totalSize
+	})
+
+	var minGroupSize int64
+	for _, g := range groups {
+		// The size of each posting list contains 4 bytes with the number of entries.
+		// We shouldn't count these as series.
+		if !g.isSubtract && !(len(g.keys) == 1 && g.keys[0] == allPostingsKey) {
+			minGroupSize = g.totalSize - int64(len(g.keys)*4)
+			break
+		}
+	}
+
+	if minGroupSize == 0 {
+		// This should also cover the case of all postings group. all postings is requested only when there is no
+		// additive group.
+		return groups, nil
+	}
+
+	var (
+		selectedSize               int64
+		atLeastOneAdditiveSelected bool
+		maxSelectedSize            = minGroupSize * seriesBytesPerPostingByte
+	)
+	for i, g := range groups {
+		postingListSize := int64(float64(g.totalSize) * s.postingListActualSizeFactor)
+		if atLeastOneAdditiveSelected && selectedSize+postingListSize > maxSelectedSize {
+			return groups[:i], groups[i:]
+		}
+		selectedSize += postingListSize
+		atLeastOneAdditiveSelected = atLeastOneAdditiveSelected || !g.isSubtract
+	}
+	return groups, nil
+}
+
+// speculativeFetchedDataStrategy selects postings lists in a very similar way to worstCaseFetchedDataStrategy,
+// except it speculates on the size of the actual series after intersecting the selected posting lists.
+// Right now it assumes that each posting list will
```

Comment:
Ehm... "will"...? :)

Reply:
🤦 I now also added a simple example to `worstCaseFetchedDataStrategy` and `speculativeFetchedDataStrategy`